### DISCUSSION TITLE
Are there any methods to arrange order of variables evaluation on locals block

### DISCUSSION THREAD

**Discussion by Arpo201** [gruntwork-io/terragrunt#3707] [Q&A] [answered]
I want to generate a new file, then read it. 
- Have cache (`terragrunt init`): Sometime it pass, sometime it fail. 
- No cache (`terragrunt init`): it always pass, because it download terraform module and script is run finished before evaluation.
- before_hook: it always fail, and it is not actually before.
```hcl
tg_render = run_cmd("tg-render.bash")
tg_config = jsondecode(file(local.tg_render_path)) # run_cmd("cat", "${local.tg_render_path}") is not work too
```

**Fail**
```log
10:58:32.377 DEBUG  Evaluated 2 locals (remaining 8): env_vars, tg_render_path
10:58:32.378 ERROR  Error: Invalid function argument
10:58:32.378 ERROR    on ../../../../../../_templates/resources/aws/s3/s3-common.hcl line 19, in locals:
10:58:32.378 ERROR    19:   tg_config      = jsondecode(file(local.tg_render_path))
```

**Pass**
```log
11:01:12.222 DEBUG  Evaluated 2 locals (remaining 8): tg_render_path, env_vars
11:01:12.223 DEBUG  run_cmd, cached output: [../../../../../../../../..
11:01:12.223 DEBUG  Evaluated 5 locals (remaining 3): tg_config, tg_render, kms_key_arn, bucket_name, aws_admin_roles
```

So, are there any methods to arrange order of variables evaluation ?

**Comment by yhakbar** [comment]
Hey @Arpo201 , why are you trying to create the file then read it? You can simply return the JSON payload as stdout from `tg-render.bash`.

HCL functions are evaluated concurrently in order to maximize performance, so unless there's a dependency between the locals, they'll be evaluated at the same time.

**Reply by Arpo201** [reply]
Thank you for your advice @yhakbar. Now, I'm using stdout from tg-render.bash, It works. Anyway, I also want to show logs on terminal for see information or troubleshooting if have some errors (Some part of tg-render.bash throw logs. I don't want to including these logs to local.variable).

**Reply by yhakbar** [reply]
You can log those messages to stderr, and they won't be returned to the local.

**Reply by Arpo201** [reply, marked as answer]
Wow, that works!! Thank you.

For anyone, this is my solution (File descriptors):

```bash
# tg-render.bash
terragrunt render-json > result.json
# Any subsequent commands that have stdout should use "1>&2"
echo "xxxx" 1>&2
cat "result.json"
```

```hcl
# terragrunt.hcl
tg_render = run_cmd("tg-render.bash")
```
Explanation:
`0` refers to standard input (stdin).
`1` refers to standard output (stdout).
`2` refers to standard error (stderr).
